src/plugins/telemetry/public/plugin.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { Plugin, CoreStart, CoreSetup, HttpStart } from '../../../core/public';
+
+import { TelemetrySender, TelemetryService, TelemetryNotifications } from './services';
+
+export interface TelemetryPluginSetup {
+  telemetryService: TelemetryService;
+}
+
+export interface TelemetryPluginStart {
+  telemetryService: TelemetryService;
+  telemetryNotifications: TelemetryNotifications;
+}
+
+export class TelemetryPlugin implements Plugin<TelemetryPluginSetup, TelemetryPluginStart> {
+  private telemetrySender?: TelemetrySender;
+  private telemetryNotifications?: TelemetryNotifications;
+  private telemetryService?: TelemetryService;
+
+  public setup({ http, injectedMetadata, notifications }: CoreSetup): TelemetryPluginSetup {
+    this.telemetryService = new TelemetryService({
+      http,
+      injectedMetadata,
+      notifications,
+    });
+
+    this.telemetrySender = new TelemetrySender(this.telemetryService);
+
+    return {
+      telemetryService: this.telemetryService,
+    };
+  }
+
+  public start({ injectedMetadata, http, overlays, application }: CoreStart): TelemetryPluginStart {
+    if (!this.telemetryService) {
+      throw Error('Telemetry plugin failed to initialize properly.');
+    }
+
+    const telemetryBanner = injectedMetadata.getInjectedVar('telemetryBanner') as boolean;
+    const sendUsageFrom = injectedMetadata.getInjectedVar('telemetrySendUsageFrom') as
+      | 'browser'
+      | 'server';
+
+    this.telemetryNotifications = new TelemetryNotifications({
+      overlays,
+      telemetryService: this.telemetryService,
+    });
+
+    application.currentAppId$.subscribe(appId => {
```

Comment:
If I understand correctly, the start method is called on every page refresh. Should we add a `stop` method to unsubscribe? I'm concerned that we might have a memory leak here.

Please discard this note if it's not valid.

Reply:
@afharo we do have checks to ensure we are not calling the intervals multiple times so even if there is a bug that calls start on every page change this shouldnt be an issue or a memory leak. I cant think of any scenario where we need stop to be called atm.